Enhance ErrReasonPVNotExist in volumebinding scheduler plugin

yibozhuang opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This change will make the message more clear when there
is a case of PVC(s) bound to PV(s) that no longer exists
and scheduler does not select the node due to this issue.

Without this change, the `FailedScheduling` error message would look like:

```
0/2 nodes are available: 2 pvc(s) bound to non-existent pv(s)
```

With this change, the message would look like:

```
0/2 nodes are available: 2 node(s) unavailable due to one or more pvc(s) bound to non-existent pv(s)
```

For larger clusters with many different reasons of nodes that
are not available, the current message can be very misleading for
users to think that there are many PVCs lost due to PVs deleted but
in fact it could be just a single PVC case but many nodes not selected
by the scheduler due to this case.

Signed-off By: Yibo Zhuang <yibzhuang@gmail.com>

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Enhanced error message for nodes not selected by scheduler due to pod's PersistentVolumeClaim(s) bound to PersistentVolume(s) that do not exist.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

ddebroy commented:
/ok-to-test

ddebroy commented:
/assign @saad-ali

ddebroy commented:
The fix to correctly attribute the count to nodes rather than PVCs looks good.

/lgtm

ahg-g commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -69,8 +69,8 @@ const (
-	// ErrReasonPVNotExist is used when a PVC can't find the bound persistent volumes"
-	ErrReasonPVNotExist = "pvc(s) bound to non-existent pv(s)"
+	// ErrReasonPVNotExist is used when a pod has PVC(s) but can't find the bound persistent volume(s)"
+	ErrReasonPVNotExist = "node(s) due to pvc(s) bound to non-existent pv(s)"
```
isn't this a pod level error, not a node level one? like all nodes would not be eligible if the pod's PVC was bound to non-existent PV.

yibozhuang replied:
right but [FindPodVolumes](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/volumebinding/binder.go#L133) is invoked for all nodes and the number of not available ones would be accumulated because everyone node would not be eligible since the PV is not found.

This change just cleans up the message to indicate that the node is not selected due to this reason.
The previous message of something like:

```
0/2 nodes are available: 2 pvc(s) bound to non-existent pv(s)
```
can be confusing since its not 2 PVCs bound to non-existent PVs, but rather 2 nodes not available to be scheduled due to pod's PVC(s) bound to non-existent PV(s).

yibozhuang committed:
Enhance ErrReasonPVNotExist in volumebinding scheduler plugin

This change will make the message more clear when there
is a case of PVC(s) bound to PV(s) that no longer exists
and scheduler does not select the node due to this issue.

Previous error message would look like:
0/2 nodes are available: 2 pvc(s) bound to non-existent pv(s)

Updated message looks like:
0/2 nodes are available: 2 node(s) unavailable due to one or more
pvc(s) bound to non-existent pv(s)

For larger clusters with many different reasons of nodes that
are not available, the current message can be very misleading for
users to think that there are many PVCs lost due to PVs deleted but
in fact it could be just a single PVC case but many nodes not selected
by the scheduler due to this case.

Signed-off By: Yibo Zhuang <yibzhuang@gmail.com>

saad-ali reviewed (approved):
Thanks 

/lgtm
/approve

yibozhuang commented:
/retest
